src/signatures.rs

```diff
@@ -0,0 +1,1206 @@
+        for (key_index, signing_key) in credential.keys {
+            let on_chain_credential = &signer_account_credentials
+                .get(&credential_index)
+                .ok_or(SignatureError::MissingIndicesOnChain {
+                    credential_index: credential_index.index,
+                    key_index:        key_index.0,
+                })?
+                .value;
+
+            let on_chain_keys = match on_chain_credential {
+                AccountCredentialWithoutProofs::Initial { icdv } => &icdv.cred_account.keys,
+                AccountCredentialWithoutProofs::Normal { cdv, .. } => &cdv.cred_key_info.keys,
+            };
+
+            let on_chain_public_key =
+                on_chain_keys
+                    .get(&key_index)
+                    .ok_or(SignatureError::MissingIndicesOnChain {
+                        credential_index: credential_index.index,
+                        key_index:        key_index.0,
+                    })?;
+
+            let VerifyKey::Ed25519VerifyKey(public_key) = *on_chain_public_key;
+
+            // Check that the public key in the `account_keys` map matches the public key on
+            // chain.
+            if signing_key.public() != public_key {
+                return Err(SignatureError::MismatchPublicKeyOnChain {
+                    credential_index:    credential_index.index,
+                    key_index:           key_index.0,
+                    expected_public_key: Box::new(public_key),
+                    actual_public_key:   Box::new(signing_key.public()),
+                });
+            };
+
+            // Generate the signature.
+            let signature = signing_key.sign(&message_hash);
+
+            // Check that the signature is valid to ensure
+            // that the public and private keys in the `account_keys` map match.
+            if !VerifyKey::from(public_key).verify(message_hash, &Signature::from(signature)) {
+                return Err(SignatureError::MismatchPublicPrivateKeys {
+                    credential_index: credential_index.index,
+                    key_index:        key_index.0,
+                });
+            }
+
+            // Insert the generated signature into the return map.
+            account_signatures
+                .sigs
+                .entry(credential_index.index)
+                .or_insert_with(|| CredentialSignatures {
+                    sigs: BTreeMap::new(),
+                })
+                .sigs
+                .insert(key_index.0, signature.into());
+        }
+    }
+
+    Ok(account_signatures)
+}
+
+/// Wrapper around the [`sign_as_account`] function to simplify the signature
+/// generation for single-signer accounts.
+pub async fn sign_as_single_signer_account(
+    client: v2::Client,
+    signer: AccountAddress,
+    signing_key: SigningKey,
+    message: &Message<'_>,
+    bi: BlockIdentifier,
+) -> Result<Signature, SignatureError> {
```

Comment:
**Question:** Should this return `AccountSignatures` like `sign_as_account`?

Reply:
As there is only one signature (in the case of a single signer), it seems awkward to store it in a two-level map and then later extract it from there again.